File: skills/spectral-embedding-learner/scripts/spectral_embedding_learner.py
#!/usr/bin/env python3
"""
Spectral Embedding Learner

Self-learning topological embedding with configurable gamut for optimal
spectral gap and fast mixing random walks on expander structures.
"""

import numpy as np
from dataclasses import dataclass, field
from typing import List, Tuple, Optional, Dict, Any
from math import sqrt, log, ceil
import random
import hashlib


@dataclass
class SpectralMetrics:
    gap: float
    lambda_2: float
    mixing_time: float
    gap_efficiency: float
    is_ramanujan: bool


@dataclass 
class EdgeLearningStep:
    node: int
    edges_added: List[Tuple[int, int]]
    gap_before: float
    gap_after: float
    ramanujan_preserved: bool


class SpectralEmbeddingLearner:
    """
    Self-learning expander graph with configurable p-adic gamut.
    
    Core invariants:
    1. Ramanujan: λ₂ ≤ 2√(d-1) always preserved
    2. Mixing: τ = O(log n) from spectral gap
    3. Ultrametric: d(x,z) ≤ max(d(x,y), d(y,z))
    """
    
    def __init__(
        self, 
        seed: int = 1069, 
        gamut_prime: int = 3, 
        target_degree: int = 4,
        target_gap_efficiency: float = 0.95
    ):
        self.seed = seed
        self.p = gamut_prime
        self.d = target_degree
        self.target_efficiency = target_gap_efficiency
        
        random.seed(seed)
        np.random.seed(seed)
        
        self.embeddings: Dict[int, np.ndarray] = {}
        self.adjacency: Dict[int, set] = {}
        self.history: List[EdgeLearningStep] = []
        
    @property
    def n_vertices(self) -> int:
        return len(self.embeddings)
    
    @property
    def ramanujan_bound(self) -> float:
        return 2 * sqrt(self.d - 1)
    
    @property
    def optimal_gap(self) -> float:
        return self.d - self.ramanujan_bound
    
    def gamut_depth(self, n: Optional[int] = None) -> int:
        """Tree depth for p-adic embedding."""
        n = n or self.n_vertices
        return ceil(log(max(n, 1), self.p)) if n > 0 else 0
    
    def p_adic_valuation(self, n: int) -> int:
        """v_p(n) = largest k such that p^k divides n."""
        if n == 0:
            return float('inf')
        k = 0
        while n % self.p == 0:
            n //= self.p
            k += 1
        return k
    
    def p_adic_norm(self, n: int) -> float:
        """|n|_p = p^(-v_p(n))."""
        v = self.p_adic_valuation(n)
        return 0.0 if v == float('inf') else self.p ** (-v)
    
    def p_adic_distance(self, emb_a: np.ndarray, emb_b: np.ndarray) -> float:
        """Ultrametric distance: d_p(a,b) = max_i |a_i - b_i|_p."""
        diff = emb_a - emb_b
        scale = 2 ** 20
        diff_int = (diff * scale).astype(np.int64)
        
        norms = [self.p_adic_norm(abs(int(d))) if d != 0 else 0.0 for d in diff_int]
        return max(norms) if norms else 0.0
    
    def adjacency_matrix(self) -> np.ndarray:
        """Build adjacency matrix from edge set."""
        n = self.n_vertices
        A = np.zeros((n, n))
        for u, neighbors in self.adjacency.items():
            for v in neighbors:
                A[u, v] = 1
                A[v, u] = 1
        return A
    
    def eigenvalues(self) -> np.ndarray:
        """Sorted eigenvalues of adjacency matrix."""
        if self.n_vertices < 2:
            return np.array([0.0])
        A = self.adjacency_matrix()
        eigs = np.linalg.eigvalsh(A)
        return np.sort(eigs)[::-1]
    
    def second_eigenvalue(self) -> float:
        """λ₂ - the key spectral quantity."""
        eigs = self.eigenvalues()
        return eigs[1] if len(eigs) > 1 else 0.0
    
    def spectral_gap(self) -> float:
        """Gap = λ₁ - λ₂."""
        eigs = self.eigenvalues()
        if len(eigs) < 2:
            return 0.0
        return eigs[0] - eigs[1]
    
    def gap_efficiency(self) -> float:
        """How close to Ramanujan bound? 1.0 = optimal."""
        if self.optimal_gap <= 0:
            return 0.0
        return min(self.spectral_gap() / self.optimal_gap, 1.0)
    
    def is_ramanujan(self) -> bool:
        """Check if current graph satisfies Ramanujan property."""
        lam2 = self.second_eigenvalue()
        return lam2 <= self.ramanujan_bound + 1e-10
    
    def mixing_time(self) -> float:
        """Theoretical mixing time from spectral gap."""
        gap = self.spectral_gap()
        if gap <= 0:
            return float('inf')
        return log(max(self.n_vertices, 1)) / gap
    
    def metrics(self) -> SpectralMetrics:
        """Current spectral metrics."""
        return SpectralMetrics(
            gap=self.spectral_gap(),
            lambda_2=self.second_eigenvalue(),
            mixing_time=self.mixing_time(),
            gap_efficiency=self.gap_efficiency(),
            is_ramanujan=self.is_ramanujan()
        )
    
    def _add_edge(self, u: int, v: int):
        """Add undirected edge."""
        if u not in self.adjacency:
            self.adjacency[u] = set()
        if v not in self.adjacency:
            self.adjacency[v] = set()
        self.adjacency[u].add(v)
        self.adjacency[v].add(u)
    
    def _remove_edge(self, u: int, v: int):
        """Remove undirected edge."""
        if u in self.adjacency:
            self.adjacency[u].discard(v)
        if v in self.adjacency:
            self.adjacency[v].discard(u)
    
    def _test_edge_preserves_ramanujan(self, u: int, v: int) -> Tuple[bool, float]:
        """Test if adding edge (u,v) preserves Ramanujan property."""
        self._add_edge(u, v)
        lam2 = self.second_eigenvalue()
        preserves = lam2 <= self.ramanujan_bound + 1e-10
        self._remove_edge(u, v)
        return preserves, lam2
    
    def _padic_candidates(self, node_id: int, k: int) -> List[Tuple[int, int]]:
        """Find k nearest nodes by p-adic ultrametric."""
        if node_id not in self.embeddings:
            return []
        
        emb = self.embeddings[node_id]
        distances = []
        
        for other in self.embeddings:
            if other != node_id and other not in self.adjacency.get(node_id, set()):
                d = self.p_adic_distance(emb, self.embeddings[other])
                distances.append((other, d))
        
        distances.sort(key=lambda x: x[1])
        return [(node_id, other) for other, _ in distances[:k]]
    
    def add_node(self, embedding: np.ndarray) -> int:
        """
        Add node with embedding, learn optimal connections.
        
        This is the SELF-LEARNING component:
        - Evaluates candidate edges by p-adic proximity
        - Selects those maximizing spectral gap
        - Rejects any that would violate Ramanujan bound
        """
        node_id = self.n_vertices
        self.embeddings[node_id] = embedding
        self.adjacency[node_id] = set()
        
        if node_id == 0:
            self.history.append(EdgeLearningStep(
                node=0, edges_added=[], 
                gap_before=0, gap_after=0, 
                ramanujan_preserved=True
            ))
            return node_id
        
        gap_before = self.spectral_gap()
        candidates = self._padic_candidates(node_id, k=self.d * 3)
        edges_added = []
        
        for _ in range(min(self.d, len(candidates))):
            best_edge = None
            best_gap = -float('inf')
            
            for (u, v) in candidates:
                if v in self.adjacency.get(u, set()):
                    continue
                    
                preserves, lam2 = self._test_edge_preserves_ramanujan(u, v)
                if preserves:
                    self._add_edge(u, v)
                    gap = self.spectral_gap()
                    self._remove_edge(u, v)
                    
                    if gap > best_gap:
                        best_gap = gap
                        best_edge = (u, v)
            
            if best_edge:
                self._add_edge(*best_edge)
                edges_added.append(best_edge)
                candidates = [(u, v) for (u, v) in candidates if (u, v) != best_edge]
        
        self.history.append(EdgeLearningStep(
            node=node_id,
            edges_added=edges_added,
            gap_before=gap_before,
            gap_after=self.spectral_gap(),
            ramanujan_preserved=self.is_ramanujan()
        ))
        
        return node_id
    
    def random_walk(self, steps: int, start: Optional[int] = None, teleport_prob: float = 0.15) -> List[int]:
        """Ergodic random walk with PageRank teleportation."""
        if self.n_vertices == 0:
            return []
        
        if start is None:
            start = random.randint(0, self.n_vertices - 1)
        
        current = start
        path = [current]
        
        for _ in range(steps):
            if random.random() < teleport_prob:
                current = random.randint(0, self.n_vertices - 1)
            else:
                neighbors = list(self.adjacency.get(current, set()))
                if neighbors:
                    current = random.choice(neighbors)
            path.append(current)
        
        return path
    
    def walk_coverage(self, steps: int, runs: int = 10) -> float:
        """Average coverage over multiple random walks."""
        coverages = []
        for _ in range(runs):
            path = self.random_walk(steps)
            coverage = len(set(path)) / max(self.n_vertices, 1)
            coverages.append(coverage)
        return np.mean(coverages)
    
    def gf3_trit(self) -> int:
        """GF(3) trit assignment based on spectral properties."""
        eff = self.gap_efficiency()
        if eff >= 0.9:
            return 0  # ERGODIC - well-balanced
        elif eff >= 0.5:
            return 1  # PLUS - generating
        else:
            return -1  # MINUS - needs validation


def demo():
    """Demonstrate spectral embedding learner."""
    print("=== Spectral Embedding Learner Demo ===\n")
    
    learner = SpectralEmbeddingLearner(seed=1069, gamut_prime=3, target_degree=4)
    print(f"Seed: {learner.seed}")
    print(f"Gamut prime: {learner.p} (depth = log_{learner.p}(n))")
    print(f"Target degree: {learner.d}")
    print(f"Ramanujan bound: λ₂ ≤ {learner.ramanujan_bound:.4f}")
    print(f"Optimal gap: {learner.optimal_gap:.4f}\n")
    
    n_nodes = 20
    dim = 64
    
    print(f"Adding {n_nodes} nodes with {dim}-dim embeddings...\n")
    
    for i in range(n_nodes):
        emb = np.random.randn(dim)
        emb = emb / np.linalg.norm(emb)
        learner.add_node(emb)
        
        if (i + 1) % 5 == 0:
            m = learner.metrics()
            print(f"  n={i+1:2d}: gap={m.gap:.4f}, λ₂={m.lambda_2:.4f}, "
                  f"eff={m.gap_efficiency:.1%}, ramanujan={m.is_ramanujan}")
    
    print(f"\n=== Final Metrics ===")
    m = learner.metrics()
    print(f"Spectral gap: {m.gap:.4f}")
    print(f"λ₂: {m.lambda_2:.4f} (bound: {learner.ramanujan_bound:.4f})")
    print(f"Gap efficiency: {m.gap_efficiency:.1%}")
    print(f"Is Ramanujan: {m.is_ramanujan}")
    print(f"Mixing time: {m.mixing_time:.2f} steps")
    print(f"GF(3) trit: {learner.gf3_trit()}")
    
    print(f"\n=== Random Walk Coverage ===")
    for steps in [10, 50, 100]:
        coverage = learner.walk_coverage(steps, runs=20)
        print(f"  {steps:3d} steps: {coverage:.1%} coverage")
    
    print(f"\n=== Learning History Summary ===")
    total_edges = sum(len(h.edges_added) for h in learner.history)
    ramanujan_violations = sum(1 for h in learner.history if not h.ramanujan_preserved)
    print(f"Total edges added: {total_edges}")
    print(f"Ramanujan violations: {ramanujan_violations}")
    
    return learner


if __name__ == "__main__":
    demo()

File: skills/spectral-embedding-learner/scripts/spectral_embedding_learner.jl
#!/usr/bin/env julia
"""
Spectral Embedding Learner - Julia Implementation with Gay.jl Integration

Self-learning topological embedding with configurable gamut for optimal
spectral gap and fast mixing random walks on expander structures.
"""

using LinearAlgebra
using Random

# Gay.jl-compatible color generation
module GayCompat
    using Random
    
    const PHI = (1 + sqrt(5)) / 2
    const GOLDEN_ANGLE = 360.0 / (PHI^2)  # ~137.508°
    
    mutable struct SplitMix64
        state::UInt64
    end
    
    function next!(rng::SplitMix64)::UInt64
        rng.state += 0x9e3779b97f4a7c15
        z = rng.state
        z = (z ⊻ (z >> 30)) * 0xbf58476d1ce4e5b9
        z = (z ⊻ (z >> 27)) * 0x94d049bb133111eb
        return z ⊻ (z >> 31)
    end
    
    function split(rng::SplitMix64)::Tuple{SplitMix64, SplitMix64}
        s1 = next!(rng)
        s2 = next!(rng)
        return (SplitMix64(s1), SplitMix64(s2))
    end
    
    function hue_at(seed::Int, idx::Int)::Float64
        rng = SplitMix64(UInt64(seed))
        for _ in 1:idx
            next!(rng)
        end
        return mod(next!(rng) / typemax(UInt64) * 360 + idx * GOLDEN_ANGLE, 360)
    end
    
    function trit_at(seed::Int, idx::Int)::Int
        rng = SplitMix64(UInt64(seed))
        for _ in 1:idx
            next!(rng)
        end
        return mod(Int(next!(rng)), 3) - 1  # -1, 0, +1
    end
    
    function hex_at(seed::Int, idx::Int)::String
        h = hue_at(seed, idx)
        # HSL to RGB (S=0.7, L=0.55)
        s, l = 0.7, 0.55
        c = (1 - abs(2l - 1)) * s
        x = c * (1 - abs(mod(h/60, 2) - 1))
        m = l - c/2
        
        r, g, b = if h < 60
            (c, x, 0.0)
        elseif h < 120
            (x, c, 0.0)
        elseif h < 180
            (0.0, c, x)
        elseif h < 240
            (0.0, x, c)
        elseif h < 300
            (x, 0.0, c)
        else
            (c, 0.0, x)
        end
        
        ri = round(Int, (r + m) * 255)
        gi = round(Int, (g + m) * 255)
        bi = round(Int, (b + m) * 255)
        
        return "#" * uppercase(string(ri, base=16, pad=2) * 
                               string(gi, base=16, pad=2) * 
                               string(bi, base=16, pad=2))
    end
end

struct SpectralMetrics
    gap::Float64
    lambda_2::Float64
    mixing_time::Float64
    gap_efficiency::Float64
    is_ramanujan::Bool
end

struct EdgeLearningStep
    node::Int
    edges_added::Vector{Tuple{Int,Int}}
    gap_before::Float64
    gap_after::Float64
    ramanujan_preserved::Bool
end

mutable struct SpectralEmbeddingLearner
    seed::Int
    p::Int  # gamut prime
    d::Int  # target degree
    target_efficiency::Float64
    
    embeddings::Dict{Int, Vector{Float64}}
    adjacency::Dict{Int, Set{Int}}
    history::Vector{EdgeLearningStep}
    
    function SpectralEmbeddingLearner(;
        seed::Int = 1069,
        gamut_prime::Int = 3,
        target_degree::Int = 4,
        target_efficiency::Float64 = 0.95
    )
        Random.seed!(seed)
        new(seed, gamut_prime, target_degree, target_efficiency,
            Dict{Int, Vector{Float64}}(),
            Dict{Int, Set{Int}}(),
            EdgeLearningStep[])
    end
end

n_vertices(L::SpectralEmbeddingLearner) = length(L.embeddings)
ramanujan_bound(L::SpectralEmbeddingLearner) = 2 * sqrt(L.d - 1)
optimal_gap(L::SpectralEmbeddingLearner) = L.d - ramanujan_bound(L)

function gamut_depth(L::SpectralEmbeddingLearner, n::Int = n_vertices(L))
    n > 0 ? ceil(Int, log(L.p, n)) : 0
end

function p_adic_valuation(n::Int, p::Int)
    n == 0 && return typemax(Int)
    k = 0
    while n % p == 0
        n ÷= p
        k += 1
    end
    return k
end

function p_adic_norm(n::Int, p::Int)
    v = p_adic_valuation(n, p)
    v == typemax(Int) ? 0.0 : Float64(p)^(-v)
end

function p_adic_distance(L::SpectralEmbeddingLearner, a::Vector{Float64}, b::Vector{Float64})
    scale = 2^20
    diff_int = round.(Int, (a .- b) .* scale)
    norms = [d != 0 ? p_adic_norm(abs(d), L.p) : 0.0 for d in diff_int]
    return maximum(norms; init=0.0)
end

function adjacency_matrix(L::SpectralEmbeddingLearner)
    n = n_vertices(L)
    A = zeros(n, n)
    for (u, neighbors) in L.adjacency
        for v in neighbors
            A[u+1, v+1] = 1.0
            A[v+1, u+1] = 1.0
        end
    end
    return A
end

function eigenvalues(L::SpectralEmbeddingLearner)
    n = n_vertices(L)
    n < 2 && return [0.0]
    A = adjacency_matrix(L)
    eigs = eigvals(Symmetric(A))
    return sort(eigs, rev=true)
end

second_eigenvalue(L::SpectralEmbeddingLearner) = begin
    eigs = eigenvalues(L)
    length(eigs) > 1 ? eigs[2] : 0.0
end

function spectral_gap(L::SpectralEmbeddingLearner)
    eigs = eigenvalues(L)
    length(eigs) < 2 ? 0.0 : eigs[1] - eigs[2]
end

function gap_efficiency(L::SpectralEmbeddingLearner)
    opt = optimal_gap(L)
    opt <= 0 ? 0.0 : min(spectral_gap(L) / opt, 1.0)
end

is_ramanujan(L::SpectralEmbeddingLearner) = second_eigenvalue(L) <= ramanujan_bound(L) + 1e-10

function mixing_time(L::SpectralEmbeddingLearner)
    gap = spectral_gap(L)
    gap <= 0 ? Inf : log(max(n_vertices(L), 1)) / gap
end

function metrics(L::SpectralEmbeddingLearner)
    SpectralMetrics(
        spectral_gap(L),
        second_eigenvalue(L),
        mixing_time(L),
        gap_efficiency(L),
        is_ramanujan(L)
    )
end

function add_edge!(L::SpectralEmbeddingLearner, u::Int, v::Int)
    haskey(L.adjacency, u) || (L.adjacency[u] = Set{Int}())
    haskey(L.adjacency, v) || (L.adjacency[v] = Set{Int}())
    push!(L.adjacency[u], v)
    push!(L.adjacency[v], u)
end

function remove_edge!(L::SpectralEmbeddingLearner, u::Int, v::Int)
    haskey(L.adjacency, u) && delete!(L.adjacency[u], v)
    haskey(L.adjacency, v) && delete!(L.adjacency[v], u)
end

function test_edge_preserves_ramanujan(L::SpectralEmbeddingLearner, u::Int, v::Int)
    add_edge!(L, u, v)
    lam2 = second_eigenvalue(L)
    preserves = lam2 <= ramanujan_bound(L) + 1e-10
    remove_edge!(L, u, v)
    return preserves, lam2
end

function padic_candidates(L::SpectralEmbeddingLearner, node_id::Int, k::Int)
    !haskey(L.embeddings, node_id) && return Tuple{Int,Int}[]
    
    emb = L.embeddings[node_id]
    distances = Tuple{Int, Float64}[]
    
    for other in keys(L.embeddings)
        other == node_id && continue
        other in get(L.adjacency, node_id, Set{Int}()) && continue
        d = p_adic_distance(L, emb, L.embeddings[other])
        push!(distances, (other, d))
    end
    
    sort!(distances, by=x->x[2])
    return [(node_id, other) for (other, _) in distances[1:min(k, length(distances))]]
end

function add_node!(L::SpectralEmbeddingLearner, embedding::Vector{Float64})
    node_id = n_vertices(L)
    L.embeddings[node_id] = embedding
    L.adjacency[node_id] = Set{Int}()
    
    if node_id == 0
        push!(L.history, EdgeLearningStep(0, [], 0.0, 0.0, true))
        return node_id
    end
    
    gap_before = spectral_gap(L)
    candidates = padic_candidates(L, node_id, L.d * 3)
    edges_added = Tuple{Int,Int}[]
    
    for _ in 1:min(L.d, length(candidates))
        best_edge = nothing
        best_gap = -Inf
        
        for (u, v) in candidates
            v in get(L.adjacency, u, Set{Int}()) && continue
            
            preserves, _ = test_edge_preserves_ramanujan(L, u, v)
            if preserves
                add_edge!(L, u, v)
                gap = spectral_gap(L)
                remove_edge!(L, u, v)
                
                if gap > best_gap
                    best_gap = gap
                    best_edge = (u, v)
                end
            end
        end
        
        if !isnothing(best_edge)
            add_edge!(L, best_edge...)
            push!(edges_added, best_edge)
            filter!(c -> c != best_edge, candidates)
        end
    end
    
    push!(L.history, EdgeLearningStep(
        node_id, edges_added, gap_before, 
        spectral_gap(L), is_ramanujan(L)
    ))
    
    return node_id
end

function random_walk(L::SpectralEmbeddingLearner, steps::Int; 
                     start::Union{Int,Nothing}=nothing, teleport_prob::Float64=0.15)
    n_vertices(L) == 0 && return Int[]
    
    current = isnothing(start) ? rand(0:n_vertices(L)-1) : start
    path = [current]
    
    for _ in 1:steps
        if rand() < teleport_prob
            current = rand(0:n_vertices(L)-1)
        else
            neighbors = collect(get(L.adjacency, current, Set{Int}()))
            !isempty(neighbors) && (current = rand(neighbors))
        end
        push!(path, current)
    end
    
    return path
end

function walk_coverage(L::SpectralEmbeddingLearner, steps::Int; runs::Int=10)
    coverages = Float64[]
    for _ in 1:runs
        path = random_walk(L, steps)
        push!(coverages, length(unique(path)) / max(n_vertices(L), 1))
    end
    return sum(coverages) / length(coverages)
end

function gf3_trit(L::SpectralEmbeddingLearner)
    eff = gap_efficiency(L)
    eff >= 0.9 ? 0 : (eff >= 0.5 ? 1 : -1)
end

function gay_color(L::SpectralEmbeddingLearner)
    GayCompat.hex_at(L.seed, n_vertices(L))
end

function demo()
    println("=== Spectral Embedding Learner (Julia + Gay.jl) ===\n")
    
    learner = SpectralEmbeddingLearner(seed=1069, gamut_prime=3, target_degree=4)
    
    println("Seed: $(learner.seed)")
    println("Gamut prime: $(learner.p) (depth = log_$(learner.p)(n))")
    println("Target degree: $(learner.d)")
    println("Ramanujan bound: λ₂ ≤ $(round(ramanujan_bound(learner), digits=4))")
    println("Optimal gap: $(round(optimal_gap(learner), digits=4))\n")
    
    n_nodes = 20
    dim = 64
    
    println("Adding $n_nodes nodes with $dim-dim embeddings...\n")
    
    for i in 1:n_nodes
        emb = randn(dim)
        emb ./= norm(emb)
        add_node!(learner, emb)
        
        if i % 5 == 0
            m = metrics(learner)
            color = gay_color(learner)
            println("  n=$(lpad(i, 2)): gap=$(round(m.gap, digits=4)), " *
                    "λ₂=$(round(m.lambda_2, digits=4)), " *
                    "eff=$(round(m.gap_efficiency*100, digits=1))%, " *
                    "ramanujan=$(m.is_ramanujan), color=$color")
        end
    end
    
    println("\n=== Final Metrics ===")
    m = metrics(learner)
    println("Spectral gap: $(round(m.gap, digits=4))")
    println("λ₂: $(round(m.lambda_2, digits=4)) (bound: $(round(ramanujan_bound(learner), digits=4)))")
    println("Gap efficiency: $(round(m.gap_efficiency*100, digits=1))%")
    println("Is Ramanujan: $(m.is_ramanujan)")
    println("Mixing time: $(round(m.mixing_time, digits=2)) steps")
    println("GF(3) trit: $(gf3_trit(learner))")
    println("Gay color: $(gay_color(learner))")
    
    println("\n=== Random Walk Coverage ===")
    for steps in [10, 50, 100]
        coverage = walk_coverage(learner, steps, runs=20)
        println("  $(lpad(steps, 3)) steps: $(round(coverage*100, digits=1))% coverage")
    end
    
    println("\n=== GF(3) Conservation Check ===")
    total_trit = gf3_trit(learner)
    println("Learner trit: $total_trit (ERGODIC=0, optimal mixing)")
    
    return learner
end

# Run if executed directly
if abspath(PROGRAM_FILE) == @__FILE__
    demo()
end
